smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/PaginationGenerator.java

```diff
@@ -0,0 +1,191 @@
+        writer.addImport(service.getName(), service.getName(), service.getNamespace());
+
+        writer.openBlock("export interface $LPaginationConfiguration extends PaginationConfiguration {",
+                "}", nonModularServiceName, () -> {
+            writer.write("client: $L | $L;", nonModularServiceName, service.getName());
+        });
+    }
+
+    private void writePager() {
+        String functionType = "export async function*";
+        writer.openBlock("$L $LPaginate(config: $L, input: $L, ...additionalArguments: any): Paginator<$L>{",
+                "}", functionType, this.methodName, this.paginationType,
+                this.inputSymbol.getName(), this.outputSymbol.getName(), () -> {
+            writer.write("let token = config.startingToken || '';");
+
+            writer.write("let hasNext = true;");
+            writer.write("let page:$L;", this.outputSymbol.getName());
+            writer.openBlock("while (hasNext) {", "}", () -> {
+                writer.write("input[\"$L\"] = token;", this.paginatedInfo.getInputTokenMember().getMemberName());
+                if (this.paginatedInfo.getPageSizeMember().isPresent()) {
+                    String pageSize = this.paginatedInfo.getPageSizeMember().get().getMemberName();
+                    writer.write("input[\"$L\"] = config.pageSize;", pageSize);
+                }
+
+                writer.openBlock("if(config.client instanceof $L) {", "}", this.nonModularServiceName, () -> {
+                    writer.write("page = await makePagedRequest(config.client, input, ...additionalArguments);");
+                });
+                writer.openBlock("else if (config.client instanceof $L) {", "}", this.serviceSymbol.getName(), () -> {
+                    writer.write(" page = await makePagedClientRequest(config.client, input, ...additionalArguments);");
+                });
+                writer.openBlock("else {", "}", () -> {
+                    writer.write(" throw new Error(\"Invalid client, expected $L | $L\");",
+                            this.nonModularServiceName, this.serviceSymbol.getName());
+                });
+
+                writer.write("yield page;");
+                if (this.paginatedInfo.getOutputTokenMember().getMemberName().contains(".")) {
+                    // Smithy allows one level indexing (ex. 'bucket.outputToken').
+                    String[] outputIndex = this.paginatedInfo.getOutputTokenMember().getMemberName().split("\\.");
+                    writer.write("token = page[\"$L\"][\"$L\"];", outputIndex[0], outputIndex[1]);
+
+                } else {
+                    writer.write("token = page[\"$L\"];", paginatedInfo.getOutputTokenMember().getMemberName());
+                }
+
+                writer.write("hasNext = !!(token);");
+            });
+            writer.write("return undefined;");
+        });
+    }
+
+
+    private void writeFullRequest() {
```

Comment:
These two methods could have clearer names and/or documentation describing what the difference is.

Reply:
Adding some documentation. Lmk if you think it needs to be adjusted